src/Middleware/RateLimiting/src/DefaultCombinedLease.cs

```diff
@@ -0,0 +1,100 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading.RateLimiting;
+
+namespace Microsoft.AspNetCore.RateLimiting;
+internal sealed class DefaultCombinedLease : RateLimitLease
+{
+    private readonly RateLimitLease? _globalLease;
+    private readonly RateLimitLease _endpointLease;
+    private HashSet<string>? _metadataNames;
+
+    public DefaultCombinedLease(RateLimitLease? globalLease, RateLimitLease endpointLease)
+    {
+        _globalLease = globalLease;
+        _endpointLease = endpointLease;
+    }
+
+    public override bool IsAcquired => true;
+
+    public override IEnumerable<string> MetadataNames
+    {
+        get
+        {
+            if (_metadataNames is null)
+            {
+                _metadataNames = new HashSet<string>();
+                if (_globalLease is not null)
+                {
+                    foreach (string metadataName in _globalLease.MetadataNames)
+                    {
+                        _metadataNames.Add(metadataName);
+                    }
+                }
+                foreach (string metadataName in _endpointLease.MetadataNames)
+                {
+                    _metadataNames.Add(metadataName);
+                }
+            }
+            return _metadataNames;
+        }
+    }
```

Comment:
Does this need to allocate everytime?

Reply:
We only allocate the first time, when `_metadataNames` is null